src/transformStateWithClones.js

```
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const history = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = Object.fromEntries(
+          Object.entries(currentState).filter(
+            ([key]) => !action.keysToRemove.includes(key),
+          ),
+        );
+        break;
+
+      default:
+        break;
```

This violates checklist item #3: 'switch/case should always have default case for error handling'. An empty `default` case will silently ignore any unknown action types. To make the function more robust, you should consider throwing an error here to signal that an unsupported action was passed.